feat(Select): Support disabling individual items

moathabuhamad-cengage opened the pull request:
Issue: #1217 

## What I did
- added support for disabling individual items in react-magma-dom
- added a unit test for that behavior 
- added an example in the docs under api/select#disabled_items

## Screenshots:
![image](https://github.com/user-attachments/assets/128a5031-22a0-4992-b23c-8f84e4497db7)


## Checklist 
- [x] changeset has been added
- [x] Pull request description is descriptive
- [x] I have made corresponding changes to the documentation
- [x] New and existing unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works

## How to test
check behavior in storybooks and in docs under api/select#disabled_items

silvalaura commented on packages/react-magma-dom/src/components/Select/shared.ts:
```diff
@@ -1,11 +1,17 @@
+import { inputBaseStyles } from '../InputBase';
 
 function buildListHoverColor(props) {
+  if (props.isDisabled) {
+    if (props.isInverse) {
+      return transparentize(0.6, props.theme.colors.neutral100);
+    }
+    return transparentize(0.4, props.theme.colors.neutral500);
+  }
```
I don't think we want to change the background/hover color. Instead, we should just change the text color similar to how we do it in Dropdown:
![image](https://github.com/user-attachments/assets/aa687e73-ae6d-46b2-b963-a30208206fbd)
![image](https://github.com/user-attachments/assets/9332acb7-c255-4210-a30e-811af1407555)

silvalaura commented:
I noticed a little bit of a strange behavior with the focus state. In the screenshot below, I was hovering over Blue, and the blue focus border kept appearing for the item below:
<img width="332" alt="image" src="https://github.com/user-attachments/assets/090ed074-1eec-4800-9a50-132fa68ab97a">

silvalaura commented on packages/react-magma-dom/src/components/Select/shared.ts:
```diff
@@ -91,19 +107,18 @@ export const StyledList = styled('ul')<{ isOpen?: boolean; maxHeight: string }>`
   line-height: 24px;
   margin: 0;
   padding: 8px 16px;
+  cursor: ${props => (props.isDisabled ? 'not-allowed' : 'pointer')};
```
@orion-cengage do you want to use the `pointer` hand or `default` which is the existing behavior?

orion-cengage replied:
I would like us to use whatever we're using on disabled buttons and other disabled items. Might be `cursor: not-allowed;`

silvalaura replied:
@orion-cengage I meant for enabled items (I agree with not allowed for disabled items). [Existing behavior](https://storybook-preview-dev--upbeat-sinoussi-f675aa.netlify.app/?path=/story/select--default) is `cursor` but this PR changes it to `pointer`

silvalaura replied:
This has not been addressed yet ^

orion-cengage replied:
Let's stick with the default. There doesn't seem to be an agreed-upon standard in other design systems. Some use the hand, some use the pointer. The native select uses the default, so let's just stick with that.

silvalaura commented:
I was able to select disabled items:
![image](https://github.com/user-attachments/assets/9b9f8b6e-2283-417d-a7d2-76959d3b83ff)
https://jam.dev/c/181a698f-8a4e-4604-a764-53c7a917b634

silvalaura commented on packages/react-magma-dom/src/components/Select/Select.stories.tsx:
```diff
@@ -153,3 +153,47 @@ LeftAlignedLabelWithContainer.decorators = [
+      { label: 'Orange', value: 'orange', disabled: false },
+      { label: 'Yellow-Disabled', value: 'Yellow-Disabled', disabled: true },
+    ]}
+    isClearable
+    initialHighlightedIndex= {2}
+    defaultSelectedItems={[
+      { label: 'Red', value: 'red' },
+      { label: 'Orange', value: 'orange', disabled: false },
+      { label: 'Yellow-Disabled', value: 'Yellow-Disabled', disabled: true },
+    ]}
+    initialSelectedItems={[
+      { label: 'Red', value: 'red' },
+      { label: 'Orange', value: 'orange', disabled: false },
+      { label: 'Yellow-Disabled', value: 'Yellow-Disabled', disabled: true },
+    ]}
```
Can these be part of the controls, so they can be changed on Storybook?

silvalaura commented on packages/react-magma-dom/src/components/Select/Select.tsx:
```diff
@@ -48,6 +48,7 @@ export function Select<T>(props: SelectProps<T>) {
     selectedItem: passedInSelectedItem,
     setReference,
     setFloating,
+    initialHighlightedIndex,
```
This is already a prop, but it's not documented. Can we properly add it to the props list with an explanation of when it's needed?

silvalaura reviewed (commented):
A change somewhere here made us lose the focus back on the select input after isClearable X is pressed like we currently have: https://storybook-preview-dev--upbeat-sinoussi-f675aa.netlify.app/?path=/story/select--default&args=isClearable:true

moathabuhamad-cengage commented:
> I was able to select disabled items: ![image](https://private-user-images.githubusercontent.com/91160746/380837115-9b9f8b6e-2283-417d-a7d2-76959d3b83ff.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3MDg1NjAsIm5iZiI6MTczMDcwODI2MCwicGF0aCI6Ii85MTE2MDc0Ni8zODA4MzcxMTUtOWI5ZjhiNmUtMjI4My00MTdkLWE3ZDItNzY5NTlkM2I4M2ZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA0VDA4MTc0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExZjFhOGU1ZjgyOTNjOWQ0NDc3NGU1NDAzMWFlN2ZlOGNhMDdjMmJmMzMzNzE1OTE3YmU0Y2YxNjRhNGEwNjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.wsLw-aogF270f-H3v2vdsBZel40A83wBh9annU8S3go) https://jam.dev/c/181a698f-8a4e-4604-a764-53c7a917b634

I wasn't able to reproduce this bug but i added more checks anyway

moathabuhamad-cengage commented:
> A change somewhere here made us lose the focus back on the select input after isClearable X is pressed like we currently have: https://storybook-preview-dev--upbeat-sinoussi-f675aa.netlify.app/?path=/story/select--default&args=isClearable:true

this seems to be broken on dev but I included a fix in this PR

silvalaura commented:
> > A change somewhere here made us lose the focus back on the select input after isClearable X is pressed like we currently have: https://storybook-preview-dev--upbeat-sinoussi-f675aa.netlify.app/?path=/story/select--default&args=isClearable:true
> 
> this seems to be broken on dev but I included a fix in this PR

This is working as expected on dev (https://next--upbeat-sinoussi-f675aa.netlify.app/api/select/), but I see that it's now been addressed in this PR as well ✅

silvalaura commented:
> > I was able to select disabled items: ![image](https://private-user-images.githubusercontent.com/91160746/380837115-9b9f8b6e-2283-417d-a7d2-76959d3b83ff.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3MDg1NjAsIm5iZiI6MTczMDcwODI2MCwicGF0aCI6Ii85MTE2MDc0Ni8zODA4MzcxMTUtOWI5ZjhiNmUtMjI4My00MTdkLWE3ZDItNzY5NTlkM2I4M2ZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA0VDA4MTc0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExZjFhOGU1ZjgyOTNjOWQ0NDc3NGU1NDAzMWFlN2ZlOGNhMDdjMmJmMzMzNzE1OTE3YmU0Y2YxNjRhNGEwNjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.wsLw-aogF270f-H3v2vdsBZel40A83wBh9annU8S3go) https://jam.dev/c/181a698f-8a4e-4604-a764-53c7a917b634
> 
> I wasn't able to reproduce this bug but i added more checks anyway

Steps to reproduce:
1. Navigate to https://storybook-preview-1493--upbeat-sinoussi-f675aa.netlify.app/?path=/story/select--with-disabled-items&args=isMulti:true
2. Click on select, then the down arrow key, then enter to select Red
3. Press the down arrow key again and notice the disabled item has focus. I expect disabled items to never get focus
4. Press arrow key down to select Green, then arrow key down to select Purple. Notice that when all that is left is disabled items, they can still gain focus

I think having `initialHighlightedIndex` is throwing things off because it's not a problem in https://storybook-preview-1493--upbeat-sinoussi-f675aa.netlify.app/?path=/story/select--multi-with-disabled-items. Although, if I change the `initialHighlightedIndex1 to 1 https://storybook-preview-1493--upbeat-sinoussi-f675aa.netlify.app/?path=/story/select--multi-with-disabled-items&args=initialHighlightedIndex:1 then `Blue-disabled` is the first item focused.

Something is off 🤔

silvalaura commented on packages/react-magma-dom/src/components/Select/index.tsx:
```diff
@@ -142,6 +142,10 @@ export interface SelectProps<T extends SelectOptions>
    */
   isMulti?: false;
+  /**
+   * Index of the item that should be highlighted by default
```
When should adopters use this?

moathabuhamad-cengage replied:
fixed

silvalaura commented on website/react-magma-docs/src/pages/api/select.mdx:
````diff
@@ -159,6 +159,71 @@ export function Example() {
+```
+
+
+```tsx
````
Instead of 2 separate sandboxes/jsx blocks for the examples, we usually just have 1 with both components we want to showcase.

moathabuhamad-cengage replied:
fixed

silvalaura commented on packages/react-magma-dom/src/components/Select/MultiSelect.tsx:
```diff
@@ -45,14 +50,53 @@ export function MultiSelect<T>(props: MultiSelectProps<T>) {
     );
   }
 
+  function isDisabledItemIndex(index) {
```
Some of these functions are repeated across MultiSelect and Select-- is it worth having them in a shared file to avoid duplication?

moathabuhamad-cengage replied:
There is no need now, using new logic for it

moathabuhamad-cengage commented:
> > > I was able to select disabled items: ![image](https://private-user-images.githubusercontent.com/91160746/380837115-9b9f8b6e-2283-417d-a7d2-76959d3b83ff.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3MDg1NjAsIm5iZiI6MTczMDcwODI2MCwicGF0aCI6Ii85MTE2MDc0Ni8zODA4MzcxMTUtOWI5ZjhiNmUtMjI4My00MTdkLWE3ZDItNzY5NTlkM2I4M2ZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA0VDA4MTc0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExZjFhOGU1ZjgyOTNjOWQ0NDc3NGU1NDAzMWFlN2ZlOGNhMDdjMmJmMzMzNzE1OTE3YmU0Y2YxNjRhNGEwNjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.wsLw-aogF270f-H3v2vdsBZel40A83wBh9annU8S3go) https://jam.dev/c/181a698f-8a4e-4604-a764-53c7a917b634
> > 
> > 
> > I wasn't able to reproduce this bug but i added more checks anyway
> 
> Steps to reproduce:
> 
> 1. Navigate to https://storybook-preview-1493--upbeat-sinoussi-f675aa.netlify.app/?path=/story/select--with-disabled-items&args=isMulti:true
> 2. Click on select, then the down arrow key, then enter to select Red
> 3. Press the down arrow key again and notice the disabled item has focus. I expect disabled items to never get focus
> 4. Press arrow key down to select Green, then arrow key down to select Purple. Notice that when all that is left is disabled items, they can still gain focus
> 
> I think having `initialHighlightedIndex` is throwing things off because it's not a problem in https://storybook-preview-1493--upbeat-sinoussi-f675aa.netlify.app/?path=/story/select--multi-with-disabled-items. Although, if I change the `initialHighlightedIndex1 to 1 https://storybook-preview-1493--upbeat-sinoussi-f675aa.netlify.app/?path=/story/select--multi-with-disabled-items&args=initialHighlightedIndex:1 then `Blue-disabled` is the first item focused.
> 
> Something is off 🤔

This was caused by mismatched indices between the filtered and unfiltered arrays, fixed by the next commit

silvalaura commented:
Issues:
1. Single select > after red is selected, if isClearable is clicked, focus goes to disabled item: https://jam.dev/c/221ef11c-aa53-4484-94ee-da3c6fd428bc
2. Not sure that `initialHighlightedIndex` is now working. I change it in storybook and I don't see it affecting the first item that gets focus

silvalaura commented on packages/react-magma-dom/src/components/Select/shared.ts:
```diff
@@ -91,19 +107,18 @@ export const StyledList = styled('ul')<{ isOpen?: boolean; maxHeight: string }>`
   line-height: 24px;
   margin: 0;
   padding: 8px 16px;
+  cursor: ${props => (props.isDisabled ? 'not-allowed' : 'default')};
```
cursor is set in line 116 and then again here, so we can clean this up
